ludwig/utils/trainer_utils.py

```diff
@@ -163,6 +171,39 @@ def log_metrics(self):
         return log_metrics
 
 
+@DeveloperAPI
+class WalltimeEarlyStopCallback(Callback):
+    """Callback that stops training when a timeout is reached and no improvement has been made."""
+
+    def __init__(self, timeout_s: float, early_stopping_steps: int = sys.maxsize) -> None:
+        """Initializes the callback.
+
+        Args:
+            timeout_s: The timeout in seconds. Can be a float to allow for sub-second timeouts.
+            early_stopping_steps: The number of steps to wait before stopping.
+        """
+        super().__init__()
+        self.timeout_delta = timedelta(seconds=timeout_s)
+        self.early_stopping_steps = early_stopping_steps
+
+    def should_early_stop(self, trainer, progress_tracker, is_coordinator):
+        """
+        Returns:
+            True if the timeout has been reached and no improvement has been made for the given early stopping steps.
+        """
+        # Time since last improvement > timeout
+        timed_out = (
+            datetime.now() - datetime.fromtimestamp(progress_tracker.last_improvement_timestamp) > self.timeout_delta
```

Comment:
Actually, something important here: how does this work with resume from previous training run? We may want to make sure this value is reset explicitly when the user is resuming from a previous training run.

Reply:
@tgaddair where does the ProgressTracker get loaded/initialized when resuming from a previous training run?

Reply:
`ProgressTracker` gets loaded in the trainer, here:

https://github.com/ludwig-ai/ludwig/blob/master/ludwig/trainers/trainer.py#L683